File: haskell/binary-search/src/BinarySearch.hs
module BinarySearch (find) where

import Data.Array

find :: (Ord a, Show a) => Array Int a -> a -> Maybe Int
find array x
  | null array = Nothing
  | otherwise = find' (minimum . indices $ array, maximum . indices $ array)
  where
    find' (l, h)
      | l > h = Nothing
      | x == middleElem = Just middleIndex
      | x < middleElem = find' (l, h - 1)
      | otherwise = find' (l + 1, h)
      where
        middleIndex = (h + l) `div` 2
        middleElem = array ! middleIndex

File: typescript/binary-search/binary-search.ts
export const find = (haystack: number[], needle: number): number => {
  const binarySearch = (from: number, to: number): number => {
    const middleIndex = Math.floor((from + to) / 2)
    const middleValue = haystack[middleIndex]

    if (middleValue === needle) return middleIndex
    if (from >= to) throw new Error('Value not in array')
    if (needle < middleValue) return binarySearch(from, middleIndex)
    if (needle > middleValue) return binarySearch(middleIndex + 1, to)
    
    throw new Error('Value not in array')
  }
  
  return binarySearch(0, haystack.length - 1)
}

